File: src/logger.ts
import { Context } from './context.ts'
import { Service } from './service.ts'
import { defineProperty, symbols } from './utils.ts'

/** Severity ranks; messages below the threshold are dropped. */
const LEVELS = { debug: 0, info: 1, success: 1, warn: 2, error: 3 } as const

type Level = keyof typeof LEVELS

function threshold(): number {
  const raw = (globalThis.process?.env ?? {}).MIN_CORDIS_LOG as Level | undefined
  return LEVELS[raw] ?? LEVELS.info
}

/**
 * Minimal named logger. Replaces the vendored Cordis logger (buffer,
 * exporters, i18n levels) with a level-filtered console sink; embedders can
 * swap `ctx.logger` for their own service since the core only calls
 * `ctx.logger.error` for error containment.
 */
export class Logger {
  /** Create a logger that prefixes every line with `name`. */
  constructor(private name: string) {}

  private output(level: Level, args: any[]) {
    if (LEVELS[level] < threshold()) return
    const sink = level === 'error' ? console.error : level === 'warn' ? console.warn : console.log
    const stamp = new Date().toISOString()
    sink(`[${stamp}] ${level.toUpperCase().padEnd(7)} ${this.name}`, ...args)
  }

  /** Log at debug severity (hidden unless `MIN_CORDIS_LOG=debug`). */
  debug(...args: any[]) { this.output('debug', args) }
  /** Log at info severity. */
  info(...args: any[]) { this.output('info', args) }
  /** Log a success marker at info severity. */
  success(...args: any[]) { this.output('success', args) }
  /** Log at warning severity. */
  warn(...args: any[]) { this.output('warn', args) }
  /** Log at error severity (stderr). */
  error(...args: any[]) { this.output('error', args) }
}

/**
 * Logger service installed as `ctx.logger`.
 *
 * Callable: `ctx.logger('my-plugin')` returns a named `Logger`. The service
 * itself also exposes level methods so `ctx.logger.error(...)` works from
 * framework containment paths.
 */
export class LoggerService extends Service {
  constructor(ctx: Context) {
    super(ctx, 'logger')
    // noShadow: the service is identity-aware (derives names from callers),
    // so tracing must keep the origin context association.
    defineProperty(this, symbols.tracker, {
      property: 'ctx',
      associate: 'logger',
      noShadow: true,
    })
  }

  /** Return a named logger bound to the service sink. */
  [Service.invoke](name = 'app'): Logger {
    return new Logger(name)
  }

  debug(...args: any[]) { this[Service.invoke]().debug(...args) }
  info(...args: any[]) { this[Service.invoke]().info(...args) }
  success(...args: any[]) { this[Service.invoke]().success(...args) }
  warn(...args: any[]) { this[Service.invoke]().warn(...args) }
  error(...args: any[]) { this[Service.invoke]().error(...args) }
}

File: python/min_cordis/_logger.py
"""Minimal named logger service, ported from the TS core's ``logger.ts``.

Replaces the vendored Cordis logger (buffer, exporters, i18n levels) with a
level-filtered console sink. The core's error containment routes to the
``on_error`` sink injected at ``Context`` construction (the Python port's
diagnostics surface); this service exposes the TS parity API —
``ctx.logger('name')`` returns a named logger, ``ctx.logger.error(...)``
records into a bounded ring for inspection.
"""

from __future__ import annotations

import os
import sys
from datetime import datetime, timezone
from typing import Any

from ._service import Service
from ._traceable import Tracker

__all__ = ["Logger", "LoggerService"]

LEVELS = {"debug": 0, "info": 1, "success": 1, "warn": 2, "error": 3}


def _threshold() -> int:
    raw = os.environ.get("MIN_CORDIS_LOG")
    return LEVELS.get(raw, LEVELS["info"])  # type: ignore[arg-type]


class Logger:
    """A named logger; every line is prefixed with ``name``."""

    def __init__(self, name: str) -> None:
        self.name = name

    def _output(self, level: str, args: tuple) -> None:
        if LEVELS[level] < _threshold():
            return
        sink = sys.stderr if level in ("error", "warn") else sys.stdout
        stamp = datetime.now(timezone.utc).isoformat()
        print(f"[{stamp}] {level.upper():<7} {self.name}", *args, file=sink)

    def debug(self, *args: Any) -> None:
        self._output("debug", args)

    def info(self, *args: Any) -> None:
        self._output("info", args)

    def success(self, *args: Any) -> None:
        self._output("success", args)

    def warn(self, *args: Any) -> None:
        self._output("warn", args)

    def error(self, *args: Any) -> None:
        self._output("error", args)


class LoggerService(Service):
    """Logger service installed as ``ctx.logger``.

    Callable: ``ctx.logger('my-plugin')`` returns a named :class:`Logger`.
    The service also exposes the level methods directly so
    ``ctx.logger.error(...)`` works from framework diagnostics paths.
    Contained errors land in the bounded ``errors`` ring (newest kept).
    """

    ERROR_RING_LIMIT = 1000

    def __init__(self, ctx: Any, config: Any = None) -> None:
        super().__init__(ctx, "logger")
        # noShadow: the service is identity-aware (derives names from
        # callers), so tracing must keep the origin context association.
        self._tracker = Tracker(property="ctx", associate="logger", no_shadow=True)
        self.errors: list[tuple] = []

    def _invoke(self, name: str = "app") -> Logger:
        return Logger(name)

    def __call__(self, name: str = "app") -> Logger:
        return self._invoke(name)

    def debug(self, *args: Any) -> None:
        self._invoke().debug(*args)

    def info(self, *args: Any) -> None:
        self._invoke().info(*args)

    def success(self, *args: Any) -> None:
        self._invoke().success(*args)

    def warn(self, *args: Any) -> None:
        self._invoke().warn(*args)

    def error(self, *args: Any) -> None:
        self.errors.append(args)
        if len(self.errors) > LoggerService.ERROR_RING_LIMIT:
            del self.errors[: len(self.errors) - LoggerService.ERROR_RING_LIMIT]
        self._invoke().error(*args)
